src/Reflection/ReflectionConstant.php

```diff
@@ -0,0 +1,283 @@
+{
+    /** @var Reflector */
+    private $reflector;
+
+    /** @var Node\Stmt\Const_|Node\Expr\FuncCall */
+    private $node;
+
+    /** @var LocatedSource */
+    private $locatedSource;
+
+    /** @var NamespaceNode|null */
+    private $declaringNamespace;
+
+    /** @var int|null */
+    private $positionInNode;
+
+    /** @var int|float|mixed[]|string|bool|null const value */
+    private $value;
+
+    /** @var bool */
+    private $valueWasCached = false;
+
+    private function __construct()
+    {
+    }
+
+    /**
+     * Create a ReflectionConstant by name, using default reflectors etc.
+     *
+     * @throws IdentifierNotFound
+     */
+    public static function createFromName(string $constantName) : self
+    {
+        return (new BetterReflection())->constantReflector()->reflect($constantName);
+    }
+
+    /**
+     * Create a reflection of a constant
+     *
+     * @internal
+     *
+     * @param Node\Stmt\Const_|Node\Expr\FuncCall $node Node has to be processed by the PhpParser\NodeVisitor\NameResolver
+     */
+    public static function createFromNode(
+        Reflector $reflector,
+        Node $node,
+        LocatedSource $locatedSource,
+        ?NamespaceNode $namespace = null,
+        ?int $positionInNode = null
+    ) : self {
+        return $node instanceof Node\Stmt\Const_
+            ? self::createFromConstNode($reflector, $node, $locatedSource, $namespace, $positionInNode)
+            : self::createFromFuncCallNode($reflector, $node, $locatedSource, $namespace);
+    }
+
+    private static function createFromConstNode(
+        Reflector $reflector,
+        Node\Stmt\Const_ $node,
+        LocatedSource $locatedSource,
+        ?NamespaceNode $namespace,
+        int $positionInNode
+    ) : self {
+        $constant                     = new self();
+        $constant->reflector          = $reflector;
+        $constant->node               = $node;
+        $constant->locatedSource      = $locatedSource;
+        $constant->declaringNamespace = $namespace;
+        $constant->positionInNode     = $positionInNode;
+
+        return $constant;
+    }
+
+    private static function createFromFuncCallNode(
+        Reflector $reflector,
+        Node\Expr\FuncCall $node,
+        LocatedSource $locatedSource,
+        ?NamespaceNode $namespace
+    ) : self {
+        $constant                     = new self();
+        $constant->reflector          = $reflector;
+        $constant->node               = $node;
+        $constant->locatedSource      = $locatedSource;
+        $constant->declaringNamespace = $namespace;
+
+        return $constant;
+    }
+
+    /**
+     * Get the "short" name of the constant (e.g. for A\B\FOO, this will return
+     * "FOO").
+     */
+    public function getShortName() : string
+    {
+        if ($this->node instanceof Node\Expr\FuncCall) {
+            /** @psalm-suppress UndefinedPropertyFetch */
+            return $this->node->args[0]->value->value;
```

Comment:
Do we know that `args[0]` is a constant? This should likely be a constant expression evaluation instead (which may fail)

Reply:
Bails out on:

```
<?php

namespace Acme;

const FOO_CONST = 'BAR';
```

because `$this->node->args[0]->value` is an instance of `PhpParser\Node\Expr\ConstFetch` which doesn't have a `$value` property.

Reply:
Strange, this should not match condition on line 116

Reply:
Ran PHP-Scoper with:

```php
$newNode = $this->node->args[0]->value;

if ($newNode instanceof Node\Scalar\String_) {
    // With:
    // define('BAR_CONST', foo());
    // $newNode->value will be 'BAR_CONST'
    //
    // However with:
    // define('Acme\BAR_CONST', foo());
    // then $newNode->value will be 'Acme\BAR_CONST'
    // and the constant declared will be Acme\BAR_CONST as in the namespace
    // Acme. Hence it's necessary to check if the value is not a namespaced string value.
    return (new Node\Name($newNode->value))->getFirst();
}

if ($newNode instanceof Node\Expr\ConstFetch) {
    // Happens with:
    // const FOO = 'BAR';
    return $newNode->name->getFirst();
}

throw new \InvalidArgumentException('Unknown node type: '.\get_class($newNode));
```

It passed the unit tests, checking for more

Reply:
I've made some improvements.